### ISSUE TITLE
OpenHAB swallows exceptions at FeatureInstaller which makes it hard to debug the problem

### ISSUE THREAD

**Issue by marad** [openhab/openhab-core#1486]
In `FeatureInstaller` there is a try-catch block that catches eventual installation errors, but only logs the error message - not the whole exception, and stack trace:
https://github.com/openhab/openhab-core/blob/d371a3432139d241b482b557fc1c74d6b7f24264/bundles/org.openhab.core.karaf/src/main/java/org/openhab/core/karaf/internal/FeatureInstaller.java#L427-L428

It results in very unhelpful error log that is missing the root cause of the problem:

```
15:28:58.982 [ERROR] [.core.karaf.internal.FeatureInstaller] - Failed installing 'openhab-ui-homebuilder, openhab-ui-basic, openhab-ui-habpanel, openhab-ui-paper': Error:
        Error downloading mvn:org.openhab.ui.bundles/org.openhab.ui.paper/[2.5.0,2.6)
```

I've tried to [reach the community](https://community.openhab.org/t/error-while-installing/98989) for help, but that didn't help. Without the stack trace it's just blind guesses.

The whole exception must be logged instead of just the message.

**Comment by wborn** [comment]
We can log the stacktraces on debug so they'll show when you increase the log level. Though the core is frozen in OH 2.5.x, so such changes will only be available in OH 3.

**Comment by marad** [comment]
Is OH3 production-ready or is it some kind of early access?

**Comment by wborn** [comment]
It's for developers to test new functionality and make it production ready and for curious users to look around. :wink:

**Comment by marad** [comment]
So there is no way to debug the issue on OH2 and I shouldn't use OH3. Shouldn't OH2 be in some kind of maintenance mode where bugs are fixed?

**Comment by wborn** [comment]
Yes there is maintenance for add-ons but not the core. Usually the error message provides enough details so for extra details like stacktraces debug logging will suffice. Most users will probably also not be very happy with big cryptic stacktraces spamming their logs every minute if the error message already provides enough details. So you are unfortunate to have an uncommon problem while the core is frozen.

**Comment by marad** [comment]
Are there somewhere instructions on how to build OpenHAB zip so I could modify Dockerfile to run my build? I could just modify the code myself but I'm not sure how to build the distribution

**Comment by openhab-5iver** [comment]
There is no need to build the entire distribution to modify one package. Just build the package, replace it in /runtime/, and then clear the cache. Or uninstall through Karaf and then install the updated one.

**Comment by marad** [comment]
@openhab-5iver is this possible at runtime for openhab-core?

**Comment by openhab-5iver** [comment]
In the console, execute...

```
bundle:install --help
```

**Comment by marad** [comment]
> Most users will probably also not be very happy with big cryptic stacktraces spamming their logs every minute if the error message already provides enough details.

@wborn I was thinking about this and while I agree that users would not be happy with exceptions spamming their logs I strongly disagree that swallowing stacktraces is a solution, because:
- As my example shows, stacktrace is **crucial** for debugging an issue, otherwise it's just a guessing game, and it's not fun after two weeks of guessing.
- When exceptions show up user should react and fix the issue regardless of his feelings. Exceptions are a signal that something is broken and need a fix.
- Any user that is careless enough to ignore such exception can just change the log level for it's logger.
- If the issue needs a fix in OpenHAB code then user will be equally unhapy about error spamming without stacktraces, and would probably still need to just change the log level. But developers will receive very limited information. Like in this very issue. Nobody really knows what is behind this error and if it's my environment configuration problem or OpenHAB's code problem. There is no way to tell.

In short: Throwing away valuable debug information just because someone could be unhappy about it is wrong direction.

**Comment by wborn** [comment]
> crucial for debugging

> valuable debug information

That's why I propose to not throw it away but log it on debug to help with debugging.

**Comment by marad** [comment]
Oh, yeah, I may have misunderstood your message, sorry! I guess debug is fine.

**Comment by splatch** [comment]
> @openhab-5iver is this possible at runtime for openhab-core?
> Could you provide me with commands for Karaf to do this?

@marad to answer your question:
```
> la -l|grep -i core.karaf                                                                                                                                                                              
188 x Active   x  80 x 1.0.0.202005242058      x mvn:org.openhab.core.bundles/org.openhab.core.karaf/x.y.z
update 188 file:/custom-mounted-in-docker.jar
```

You can also do an overlay of docker image:
```
FROM openhab/openhab:x.y.z
ADD custom.jar /openhab/runtime/system/org/openhab/core/bundles/org.openhab.core.karaf/2.5.0/org.openhab.core.karaf-2.5.0.jar
```

Best regards,
Łukasz
